### ISSUE TITLE
width subtyping with nonexact known properties fails in idiomatic javascript 

### ISSUE THREAD

**Issue by villesau** [facebook/flow#6146]
See in [try flow](https://flow.org/try/#0JYWwDg9gTgLgBAKjgQwM5wEoFNkGN4BmUEIcA5FDvmQFA0wCeYWcAygK4BGAjAArFh0AXjgBvFAC44AO3YhOWKABo4nKbPmK4AXwDc9Jiw6cATPwiC4I8cimoYUYNIDmK3HYdPnO-TVwRpezYubis4AAowAVQpYz5ogEorAD44AB4AE2AANzgAemT9f0D4YxMwyOjYrjNElPSs3ILfYqCAYRIK0Ro4OFQsABssfGglHrgAOimoi1QabSlxfqGRqCkHdiwVKYm480ttOAAfMT7B4ZhoKQJkAf7tqbL99G0koWTx5YvoOAB+dLiYh2M0EhwKcCkaTKQOm0TBhSAA)

Documentation mentions width-subtyping, caveats and how to use exact types to be able to be more precise: https://flow.org/en/docs/lang/width-subtyping/

However, there are cases where all the possible types are known even when the ojbects are not exact.

As you can see, both object properties defines property `selector`. This means that the property can be either `true` or `false` and based on that props has at least bunch of properties. If it's `true`, we know that property `a` will be number, _not_ string etc.

Anyhow this is not working in flow which is major defect since the code pasted in try flow is pretty idiomatic way of using Javascript, and as far as I know, there can't be uncertainty of defined types which means it should be completely possible for Flow to type check the above code properly.

PS. the above example does not work with exact types either.

**Comment by apsavin** [comment]
[Here](https://flow.org/try/#0JYWwDg9gTgLgBAKjgQwM5wEoFNkGN4BmUEIcA5FDvmQFA0wCeYWcAygK4BGAjAArFh0AXjgBvFAC44AO3YhOWKABo4nKbPmK4AXwDc9Jiw6cATPwiC4I8cimoYUYNIDmK3HYdPnO-TVwRpezYubis4AAowAVQpYz5ogEorAD44AB4AE2AANzgAemT9f0D4YxMwyOjYrjNElPSs3ILfYqCAYRIKqIsYsThULAAbLHxoKQd2LBUAOlmAEgBRAA88GDS480FU7TgAHz6B4dGoKQJkQYGZ+eXV9ZrN1G2koVTRGjg4bsFpw5GYaDgAH50nExLNpl9UDsCnApHdTGDZpDoYUaNogA) the fixed example.

Issues with destructuring of unions are alredy reported: https://github.com/facebook/flow/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aopen+union+destructuring

**Comment by villesau** [comment]
@apsavin that's not logically the same. In your example, `selector` is passed into sub components when in the original example it's parsed away from the props. I can live with that, but it's suboptimal as I need to change the code in order to make it work with Flow.

**Comment by apsavin** [comment]
> it's suboptimal as I need to change the code in order to make it work with Flow.

Agree, I hope it will be fixed with the ability to use destructuring of unions.

**Comment by villesau** [comment]
This is actually very broken with exact types too: [Try Flow](https://flow.org/try/#0JYWwDg9gTgLgBAKjgQwM5wEoFNkGN4BmUEIcA5FDvmQFA0wCeYWcAygK4BGAjAArFh0AXjgBvAD4oAXHAB27EJyxQANHE4z5i5XHEBfANz0mLDpwBM-CILgiJ0uKhhRgsgOZrcMpy-e7DdLgQsk5sXNy2cAAUYAKoMmZ8cQCUtgB8cAA8ACbAAG5wAPRpRkEh8GbmkTFxCVyWKelZuQXFRjRloQDCJNWiNHCOWAA2WPjQKgNwAHSzsdaoNHoy9qgjYzDQMs7sWGqz04lWNvq6YpJro+NQMgTIw2v7s5XH6PqpQmlTlxvQcAD8WUSYgO80EeiKGRkmUqILmcQhbSAA)

There might be duplicate of this issue though.

**Comment by villesau** [comment]
Here are bunch of related issues:

https://github.com/facebook/flow/issues/6805
https://github.com/facebook/flow/issues/6594
https://github.com/facebook/flow/issues/6408
https://github.com/facebook/flow/issues/5745
https://github.com/facebook/flow/issues/4772
https://github.com/facebook/flow/issues/3932

Could e.g @mrkev or @jbrown215 comment on wether this could be Prioritized? Looks like there are quite a many developers struggling with this.

**Comment by jbrown215** [comment]
cc @samwgoldman, who has a lot of changes to object types in the works.

**Comment by samwgoldman** [comment]
There are a few issues here, one is a know bug with destructuring unions of objects. Here is a working version:

```
import * as React from 'react'

type Sub1Props = {| a: number, b: number |};
type Sub2Props = {| a: string, c: string |};

const Sub1 = (props: Sub1Props) => <div />;
const Sub2 = (props: Sub2Props) => <div />;

const Com = (props: {| selector: true, ...Sub1Props |} | {| selector: false, ...Sub2Props |}) => {
  if (props.selector) {
    let {selector, ...rest} = props;
    return <Sub1 {...rest} />;
  } else {
    let {selector, ...rest} = props;
    return <Sub2 {...rest} />;
  }
}
```

Once you destructure `selector` and `props` into separate variables, you can't refine `props` by doing a check on `selector`—the relation between these variables is not tracked by Flow. Instead, you need to keep the union together, discriminate between the arms of the union as I've done above, then you can destructure.

**Comment by shanshanzhu** [comment]
@samwgoldman  The solution is not ideal, especially if we wanna use flowtype to represent the proto3 oneof field. Is there any plan to improve it?
https://developers.google.com/protocol-buffers/docs/reference/proto3-spec#oneof_and_oneof_field
